examples/siamese_network/siamese_network.py

```diff
@@ -0,0 +1,270 @@
+        return self.data.shape[0]
+
+    def __getitem__(self, index):
+        """
+        For every sample in the batch we select 3 images. First one is the anchor image
+        which is the image obtained from the current index. We also obtain the label of
+        anchor image.
+
+        Now we select two random images, one belonging to the same class as that of the
+        anchor image (named as positive_image) and the other belonging to a different class
+        than that of the anchor image (named as negative_image). We return the anchor image,
+        positive image, negative image and anchor label.
+        """
+
+        # obtain the anchor image
+        anchor_image = self.data[index].clone().float()
+
+        # obtain the class label of the anchor image
+        anchor_label = self.dataset.targets[index]
+        anchor_label = int(anchor_label.item())
+
+        # find a label which is different from anchor_label
+        neg_label = random.randint(0, 9)
+        while neg_label == anchor_label:
+            neg_label = random.randint(0, 9)
+
+        # find the index of image with positive label
+        positive_index = random.choice(self.grouped_examples[anchor_label])
+
+        # find the index of image with negative label
+        negative_index = random.choice(self.grouped_examples[neg_label])
+
+        # retrieve the image
+        positive_image = self.data[positive_index].float()
+
+        # retrieve the image
+        negative_image = self.data[negative_index].float()
+
+        return anchor_image, positive_image, negative_image, anchor_label
+
+
+def run(args, model, device, optimizer, train_loader, test_loader, lr_scheduler):
+
+    # using Triplet Margin Loss
+    criterion = nn.TripletMarginLoss()
+    average_test_loss = 0
+
+    # define model training step
+    def train_step(engine, batch):
+        model.train()
+        anchor_image, positive_image, negative_image, anchor_label = batch
+        anchor_image = anchor_image.to(device)
+        positive_image, negative_image = positive_image.to(device), negative_image.to(device)
+        anchor_label = anchor_label.to(device)
+        optimizer.zero_grad()
+        anchor_out, positive_out, negative_out = model(anchor_image, positive_image, negative_image)
+        loss = criterion(anchor_out, positive_out, negative_out)
+        loss.backward()
+        optimizer.step()
+        return loss
+
+    # define model testing step
+    def test_step(engine, batch):
+        model.eval()
+        anchor_image, positive_image, negative_image, anchor_label = batch
+        anchor_image = anchor_image.to(device)
+        positive_image, negative_image = positive_image.to(device), negative_image.to(device)
+        anchor_label = anchor_label.to(device)
+        anchor_out, positive_out, negative_out = model(anchor_image, positive_image, negative_image)
+        test_loss = criterion(anchor_out, positive_out, negative_out)
+        return test_loss.item()
```

Comment:
Testing step and metric should be something like accuracy: given two images of the same class -> model accuracy that model correctly predicts that they belong to the same class. Same for samples from different classes.

Reply:
Yes, that will be a better measure of testing the model. In the original paper they have described as follows, "given a pair of two face images a squared L2 distance threshold D(xi, xj) is used to determine the classification of same and different." I will work on the changes!

Reply:
@vfdev-5 I read the FaceNet paper and tried tweaking the model structure and threshold values for classifying between same and different images. However, the accuracy which I am getting is not at par with the best models on CIFAR10 datasets. 
The problem with the Triplet Loss approach is that in the same category, let's say cars, there are many different types and colours of cars, hence the model is not able to `learn` the features properly. Even so, the loss function is trying to minimize the squared Euclidean distance between two images of the same category and maximize it between two images belonging to different category but there are some common features in between the images of different categories, for example `cars` and `trucks` categories are having similar features like tyres, which might the cause of poor performance. 
I believe implementing the Siamese Network and Triplet Margin Loss on the dataset used in the original FaceNet paper will be a better option.

Reply:
In the original implementation in [pytorch/examples/siamese_network](https://github.com/pytorch/examples/blob/main/siamese_network/main.py), the implementation seems wrong to me, since in the model is using a sigmoid output after passing through the network and finally the predictions are made using the threshold value of 0.5 (i.e. if the output value is greater than 0.5 then the prediction is correct, incorrect otherwise). The problem in this approach is that the output of the model is there is no classification as such and we are just checking the sigmoid passed value of the final output of the model which is coming out to be near 1 for most of the pixels and hence the increased accuracy!

Reply:
@DeepC004 Thanks for the updates !

Concerning learning on CIFAR10 dataset, we try few other options: 1) try triplet loss with another distance, https://pytorch.org/docs/stable/generated/torch.nn.TripletMarginWithDistanceLoss.html, 2) improve feature extraction part in the model. 
I agree that loss function choice is very important. However, I think it's still possible to learn on cifar10, for example ultimately a pretrained for 10 labels classification resnet18 will be able to provide "features" (one-hot encoded vector for predicted class) such that computed similarity distance would make sense.

> I believe implementing the Siamese Network and Triplet Margin Loss on the dataset used in the original FaceNet paper will be a better option.

I'm not sure about something like FaceNet dataset if it is available on torchvision easily.

> Only the sigmoid passed value of the final output of the model are being checked. These values come out to be almost 1 for most of the pixels and hence the increased accuracy!

I'm not sure to fully understand your point here. Can you give an example and also say how to fix the issue in the example.

Reply:
Thanks for the prompt response! In regards to the points raised:
1) I will try running the model with the mentioned loss function
2) I will tweak the model layers/use the more recent models which have shown better results
Regarding the loss function part, the model is not directly giving out the one hot encoded vector for the 10 classes, rather it is using the label to select an image and then using that anchor label to find positive and negative images accordingly. The model I recently tried running is similar to the one described in the original FaceNet paper (I updated it to include the FC layers in the end and changed the activation functions to ReLU). 
The LFW dataset (used in the FaceNet paper) is open source and is easily available.

My plan of action is as follows:
1) I will tweak the loss function, and the model architecture, both standalone and simultaneously, and check if I can get any better results.
2) Since LFW dataset is available, I will implement the same model on the dataset and check how it performs. Incase option 1 does not work then we can go with this example.

This piece of [code](https://github.com/pytorch/examples/blob/7ec911c46c00ac98d3adfd996ee2a011bbb9fdba/siamese_network/main.py#L70-L78) is just concatenating the anchor image and a randomly chosen positive or negative image followed by passing through a FC layer which is followed by a sigmoid layer. I am not completely sure if I understand the reason for doing this in the mentioned example.

Reply:
@DeepC004 thinking more about all these things: siamese networks and FaceNet. Seems like we are more trying to correctly implement FaceNet paper. In this case, it would be better to check most prominent repos from "paper with code": https://paperswithcode.com/paper/facenet-a-unified-embedding-for-face

In such case, this PR with an example is not just about siamese networks but FaceNet as image recognition and clustering.

I think it would be better to rescope the purpose of the example and also implement something more recent (paper by 2021, 2022, 2023).

I agree with you that pytorch/example code can be doing something different from they claim to do. I think it is correctly labelled as siamese networks example than facenet.

If we want to stick with siamese network idea, we can : 

1) try to revert this code to pytorch/example one: use BCELoss, model does concat of some resnet features as in the example and outputs similarity of 2 images. But let's still use that on CIFAR10. We should however complement the code with more relevant metrics as discussed in this PR. Maybe, also add some visualizations. 

2) let's pick something exciting from : https://paperswithcode.com/method/siamese-network

Reply:
It will be great to complete this example and then move on to a more recent implementation as a separate example. I will make the mentioned changes, test the model and update you on that. Post this I can look into other interesting papers!

Reply:
Sounds good to me, thanks !

Reply:
Hi, sorry for the late response since I had my mid terms the past week. I tried making a few different changes to the model architecture and tested them locally. Results are as follows:
1) I updated the original code taken from [pytorch/examples/siamese_network](https://github.com/pytorch/examples/blob/main/siamese_network/main.py) to make it work on CIFAR10 dataset. The model gave an accuracy of around 71%. I checked other one vs all classifier models (multi class classification using BCE loss) on CIFAR10 dataset and they have accuracy results in similar ranges.
2) Another network architecture which makes use of Triplet Margin Loss and random triplet selection gave an accuracy of around 75%. A good classification depends on appropriate choice of threshold for the pairwise distances. These results are using ResNet34 as base network.